internal/syncutil/limit.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package syncutil
+
+import (
+	"context"
+
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/sync/semaphore"
+)
+
+// LimitRegion provides a way to bound concurrent access to a code block.
+type LimitRegion struct {
+	ctx     context.Context
+	limiter *semaphore.Weighted
+	ended   bool
+}
+
+// NewLimitRegion creates a new LimitRegion.
+func NewLimitRegion(ctx context.Context, limiter *semaphore.Weighted) *LimitRegion {
```

Comment:
Let's follow the name of `LimitedReader` and `LimitReader()` in the package `io`.

Then we have
- `LimitRegion` --> `LimitedRegion`
- `NewLimitRegion` --> `LimitRegion`

Reply:
Sounds good.